Maya Viewport 2.0 rendering for SceneShape

mattigruener opened the pull request:
I'm PR'ing this even though some of the issues we already discussed still persist. Probably worth talking about anyway.

John, I think I've addressed your early feedback and hopefully made things a little clearer... I've also taken Don's feedback as motivation to restructure lots of the code to be more in line with what Maya seems to expect us to do and to clean things up. The feedback from both of you has made this code a lot better, thanks for your help so far.

edit: Ignore the commit that has the Timer changes. I need to look into fixing up that other PR that deals with the Timer bits and currently just have this here for convenience.

donboie commented:
Hey Matti, 

Subject to the agreement of @andrewkaufman or @johnhaddon, I think you should address function renames, addition of a few more comments only for now. My argument is we should just merge this now as it's huge improvement over what we have  and there isn't an API which needs be supported. 

I think there a number of future improvements we can consider: 

- using some other container to store scene location, time, render style and RenderItem name and use this  instead of encoding this information in the RenderItem name.
- sharing buffers - perhaps a few special hidden RenderItems to store these arrays?
- GPU memory cost estimate
- Rendering tests - should be simple to implement in this case and useful when we ugrade maya or refactor this code.

Great job!

andrewkaufman commented on include/IECore/Timer.h:
```diff
@@ -92,13 +93,25 @@ class IECORE_API Timer
 
 class IECORE_API ScopedTimer
 {
+	typedef std::string Channel;
```
Can we drop this whole commit from the PR? Its unrelated and has its own PR with review comments already.

mattigruener replied:
Yes sorry, I had edited the description of the PR after that initial email got sent. It's just here still in case one of your comments requires me to do some profiling.

andrewkaufman commented on src/IECoreMaya/SceneShapeInterface.cpp:
```diff
@@ -1094,6 +1094,18 @@ MPxSurfaceShape::MatchResult SceneShapeInterface::matchComponent( const MSelecti
 	return MPxSurfaceShape::matchComponent( item, spec, list );
 }
 
+MSelectionMask SceneShapeInterface::getShapeSelectionMask() const
```
Am I to understand this is working again?

mattigruener replied:
Nope... But these bits are definitely gonna be needed, I think.

andrewkaufman commented on python/IECoreMaya/SceneShapeUI.py:
```diff
@@ -383,9 +383,11 @@ def __selectedSceneShapes() :
+	maya.cmds.selectType( objectComponent=True, allComponents=False, facet=True )
 	maya.cmds.hilite( parent )
 
+	maya.cmds.select( clear=True )
```
Out of curiosity, what bug did this fix?

mattigruener replied:
Good question - I remember that I had to do this, probably in order to not mess up the state in which the wireframes are drawn. I will go back and see if this is needed once we have selection working again.

johnhaddon commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	}
+}
+
+MRenderItem *getRenderItem( MSubSceneContainer &container, ConstObjectPtr object, const MString &name, RenderStyle style, bool &isEmpty )
```
> Perhaps createOrAcquireRenderItem...

For what it's worth, in Gaffer `create()` methods always return a new instance, and `acquire()` methods either create one if it doesn't exist or reuse one that exists already. So in Gaffer parlance this would just be `acquireRenderItem()`...

mattigruener replied:
Addressed in cdc2c2a. Using gaffer lingo now.

johnhaddon commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+
+MRenderItem *getRenderItem( MSubSceneContainer &container, ConstObjectPtr object, const MString &name, RenderStyle style, bool &isEmpty )
+{
+	MRenderItem *renderItem = container.find( name );
```
I notice that this function looks in the container to see if it can reuse an item, but then expects the caller to deal with putting new items back into the container to be reused next time. That splitting of responsibility seems very error prone, particularly given the extra details of `isEmpty` (that I don't understand). Is it possible to add the new items to the container in this function instead, so all responsibility is located in one place?

mattigruener replied:
Addressed in 97bd2ef. It is possible to do it that way, afaik. I had a version of all this where selection was working and I've really just pulled out the `container.add()` to exactly replicate the order in which I was creating the MRenderItem, assigning a shader, putting it in the container and then filling it with geometry buffers. That didn't help though and I've moved it back now. Thanks for catching that!

johnhaddon commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	return result;
+}
+
+bool sceneIsAnimated( ConstSceneInterfacePtr sceneInterface )
```
Pass raw pointers (`const SceneInterface *`) when a function does not need to retain ownership. This has less overhead and also expresses the function's semantics better.

mattigruener replied:
Addressed in 1ec828b

johnhaddon commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+
+bool sceneIsAnimated( ConstSceneInterfacePtr sceneInterface )
+{
+	ConstSampledSceneInterfacePtr scene = IECore::runTimeCast< const SampledSceneInterface >( sceneInterface );
```
Likewise, don't increment the reference count here, just use a raw pointer.

mattigruener replied:
Addressed in 1ec828b

johnhaddon commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+		// \todo: There's gotta be a better way...
+		Imath::Box3d boundingBoxD = sceneInterface->readBound( m_time );
+		Imath::Box3f boundingBox( Imath::V3f( boundingBoxD.min.x, boundingBoxD.min.y, boundingBoxD.min.z ), Imath::V3f( boundingBoxD.max.x, boundingBoxD.max.y, boundingBoxD.max.z ) );
+		const MBoundingBox mayaBoundingBox = IECore::convert<MBoundingBox>( boundingBox ); //\todo: can this be constructed from Box3d?
```
The "IECoreMaya/Convert.h" header suggests it can - is that not working?

mattigruener replied:
Ah yes, it is of course. I think that comment slipped through and my main gripe was with converting between Box3f and Box3d. I've addressed all of this in d48d761 and the double->float conversion happens down the road now.

mattigruener commented:
Thanks for the feedback, guys!

One thing I haven't mentioned - currently we're not using the instanced rendering capabilities of VP2 because I ran into those issues of geometry disappearing. I can look into adding that stuff back in in one of the next iterations on this. 
What that means, though, is that we don't get as much out of the viewport as we could... One of my test scenes consists of a heavily instanced environment (think lots of rocks) and opening that cache and tumbling was possible at about 2fps when using instancing. Now, without instancing, Maya won't even open that file anymore. I think there is still a lot of performance that we can squeeze out in general.

While talking about future improvements - we should look into supporting instancers. Especially with people getting into MASH this would be really nice to have.

I'll address your feedback either by replying to your comments or by putting up a commit with a fix. Thanks again.

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+
+void fillCurvesData( ConstObjectPtr object, GeometryDataPtr geometryData )
+{
+	// \todo: currently a curve is built from linear segments. Needs proper interpolation at some point.
```
That's correct. I've tested with a few files and found the visual difference to be fairly small, but yeah, it's definitely ignoring the basis at the moment.

andrewkaufman replied:
I'm happy to leave it until we get user feedback, but here are some example curves that I expect will look quite odd if we just draw linear:

```
/data/share/andrewk/curve.cob
/data/share/andrewk/curves.cob
```

mattigruener replied:
Oh yes, quite odd indeed.

andrewkaufman replied:
@murrays-ie and company didn't seem concerned about this for a first pass, since our grooms usually have enough cvs for linear not to be a distraction.

mattigruener replied:
Great, thanks for checking with them!

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	return result;
+}
+
+Imath::M44d worldTransform( const IECoreScene::SceneInterface *scene, double time )
```
I'm pretty sure I've stolen this exact code from somewhere else, yes. I'm inclined to throw this away completely and just accumulate the transform while walking the hierarchy. That's what I did before lots of debugging regarding disappearing objects and malfunctioning selection, which caused me to question my transforms and bounding boxes. Using this function outside of the general recursion made things a bit easier in terms of debugging.

mattigruener replied:
Addressed in ed00b8e.

I've switched back to accumulating the matrices while traversing the hierarchy and removed that duplicated function entirely.

mattigruener commented on include/IECoreMaya/SceneShapeSubSceneOverride.h:
```diff
@@ -0,0 +1,154 @@
+		using Instances = std::vector<Instance>;
+
+		// Traverse the scene and create MRenderItems as necessary while collecting all matrices to be associated with them.
+		void visitSceneLocations( IECoreScene::ConstSceneInterfacePtr sceneInterface, RenderItemMap &renderItems, MHWRender::MSubSceneContainer &container, const Imath::M44d &matrix, bool isRoot = false );
```
Addressed in 1ec828b

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	}
+}
+
+void SceneShapeSubSceneOverride::gatherInstanceInformation( Instances &instances ) const
```
Addressed in cc03db1

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+			MShaderInstance *shader = getShader( m_sceneShape->thisMObject(), RenderStyle::BoundingBox, instance.componentMode, /* isComponentSelected = */ false );
+			renderItem->setShader( shader );
+
+			container.add( renderItem );
```
Addressed in 97bd2ef

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+			MShaderInstance *shader = getShader( m_sceneShape->thisMObject(), style, instance.componentMode, instance.componentMode ? componentSelected : instance.selected );
+			renderItem->setShader( shader );
+
+			container.add( renderItem );
```
Addressed in 97bd2ef

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	}
+}
+
+void getMBuffersForRenderItem( GeometryDataPtr geometryData, const MVertexBufferDescriptorList &descriptorList, RenderStyle style, MVertexBufferArray &vertexBufferArray, MIndexBuffer &indexBuffer )
```
Addressed in 88062f7

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+		return;
+	}
+
+	IECore::ConstObjectPtr object = sceneInterface->readObject( m_time );
```
Addressed in c256bfb

mattigruener commented on include/IECoreMaya/SceneShapeSubSceneOverride.h:
```diff
@@ -0,0 +1,154 @@
+namespace IECoreMaya
+{
+
+enum class RenderStyle { BoundingBox, Wireframe, Solid, Textured, Last };
```
Addressed in 76a090e.

I shuffled bits around in order to make these fairly invisible to the outside. Could you have a quick look at what I've done?

andrewkaufman replied:
lgtm

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+
+bool isPathSelected( const MSelectionList &selectionList, const MDagPath &path )
+{
+    MStatus status = MStatus::kSuccess;
```
Addressed in 9f6a21f

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	}
+}
+
+RenderStyle& operator++( RenderStyle& style )
```
I've removed it here: 163b37e. Does that look reasonable, John?

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	IECore::IntVectorDataPtr wireframeIndexData( new IECore::IntVectorData() );
+	std::vector<int> &indexWritable = wireframeIndexData->writable();
+
+	indexWritable.push_back( 0 );
```
Thanks, that's nicer. Changes in 358c5cf

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+
+// Cache for geometric data
+using GeometryDataCache = IECore::LRUCache<IECore::MurmurHash, GeometryDataPtr, IECore::LRUCachePolicy::Parallel, GeometryDataCacheGetterKey>;
+GeometryDataCache g_geometryDataCache(geometryGetter, 100);
```
I addressed this in 48f2022. 

I wasn't sure about the defaultObjectPool() solution as that would mean that we'd have to expose the GeometryData implementation more formally, - deriving from Object and giving it a TypeId? Is that correct? Also, are we concerned about interfering/slowing down other systems that use that cache? I've used a separate limit for that one for now. 

Note for future reference: Maya's MRenderer class provides access to `GPUtotalMemorySize()` and `GPUUsedMemorySize()`. Might come in handy at some point.

andrewkaufman replied:
Yeah, I'm not sure really. From a user perspective, its a pity we have so many different caches... its really hard for a user to say "Cortex can use N mb in Maya" right now... Minimally we'll need to re-run your various scenes to gauge performance in the different production scenarios and determine if 500mb is a reasonable limit (at IE set the defaultObjectPool to 4gb for example, and some users increase it much higher). 500 is fine for the code, we just need to know how we'll set that env var here.

@johnhaddon any thoughts on whether to make a dedicated VP2 cache or try to hook into some other existing Cortex cache?

johnhaddon replied:
Actually, I realise I don't even know why we're caching the GeometryData. That's an intermediate form isn't it? And Maya will have its own representation which is actually used for the drawing? If so, can we just cache the final form? Sorry if this is a dumb question...

mattigruener replied:
Not a dumb question, at all. 

It's an intermediate form, yes, and the final form are the `MRenderItem` objects that Maya uses for rendering. I initially played with caching just those guys, but that led to really messy code. The reason being that once we hand them off to Maya by putting them in the container Maya provides, Maya assumes ownership and takes the right to destroy them at any time (potentially once the GPU memory is filled up, and definitely also when we call `clear()` on the container). So, holding on to them is no good. My next test was then to always hand off copies to Maya, but the Maya API really isn't made for that and they really expect you to use the API in the way we do now. While you can copy MRenderItems for example, you cannot rename them and because the container indexes them by name, you can't put in a copy twice (for identical instances for example). All of that was really messy and I eventually converged to being closer to their examples which is what we're reviewing at the moment... I probably should have done that from the beginning, but maintaining the stash of objects to be rendered myself was tempting since we have the LRUCache and learning how Maya wants things done was a bit of a process.

As a side note, my hope with this now is that for the second iteration it will make sense to use the hash we get for the GeometryData as the name of the respective MRenderItems in the container. Looking up the MRenderItems and appending an instance matrix for every successful lookup should then give us instanced rendering and the performance improvements I saw from this earlier. That might me fairly low hanging fruit at this point - if the rendering works and geo isn't dropping out anymore.

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+
+	geometryData->positionData = p;
+
+	IECore::IntVectorDataPtr wireframeIndexData( new IECore::IntVectorData() );
```
Shall we compile a list of future performance improvements and put it on there, or do you think we should address this for the first version?

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	}
+
+	// \todo compute cost from geometry data
+	cost = 1;
```
Addressed in 48f2022.

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	geometryData->wireframeIndexData = wireframeIndexData;
+
+	// \todo: For completeness we should probably create data for rendering a shaded box at some point
+	geometryData->normalData.reset( new IECore::V3fVectorData() );
```
Now they can: cf69b97

mattigruener commented on include/IECoreMaya/SceneShapeSubSceneOverride.h:
```diff
@@ -0,0 +1,154 @@
+#ifndef IE_COREMAYA_SCENESHAPESUBSCENEOVERRIDE_H
+#define IE_COREMAYA_SCENESHAPESUBSCENEOVERRIDE_H
+
+#include <bitset>
```
Is this ok? 944cdee

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+#include "maya/MFnDependencyNode.h"
```
Is this ok? 944cdee

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	return false;
+}
+
+void SceneShapeSubSceneOverride::update( MSubSceneContainer& container, const MFrameContext& frameContext )
```
How are the changes in f7b47e0, Don?

mattigruener commented on src/IECoreMaya/SceneShapeSubSceneOverride.cpp:
```diff
@@ -0,0 +1,1615 @@
+	}
+}
+
+bool SceneShapeSubSceneOverride::renderAllBounds( unsigned int displayStyle ) const
```
How do you like these changes? 091ffbf

mattigruener commented:
I think I've addressed all comments, except two from Andrew and Don about the `ensureFaceVaryingData()` function. I'm not sure if we can reuse some existing code because I need to get some mapping between old and new indices to update the wireframe indices so that we can render non-triangulated wireframes. Does that make sense? @andrewkaufman?

andrewkaufman commented:
I don't remember making a comment about `ensureFaceVaryingData`... can you point me to what I said?

Also, it looks like the Timer commit is still here.

mattigruener commented:
Dropped the Timer commit now.

You're right, the comment was not about that function, but about the PrimVar handling right before we call it. I was referring to this one: "Can you use the new IndexedView instead of expanding?". Maybe we can have a quick chat about how you expect me to use it :) 

Thanks.

andrewkaufman commented:
@mattigruener do you want to rebase in all those feedback commits now? Or is that quite difficult?

Did you make the change regarding the env var logic?

What did you decide about `IndexedView`?

Was there anything else left to address?

mattigruener commented:
Shouldn't be difficult - Don is looking through it and testing it in Maya, and I was planning on updating this PR once he's done.

> Did you make the change regarding the env var logic?

Don't think so. Will add that change when I push the rebased code. Thanks for the reminder.

> What did you decide about IndexedView?

I thought I'd leave it for now so we can move on. It's easy enough to change later, I guess.

mattigruener commented:
Squashed those feedback commits and added commit that handles our env var concerns.

Missing: Apparently there was an API change between Maya2016 and 2017 that I'm not handling. Been testing with 17 and when Don tried to compile for our current Maya version, it didn't work. Will look into that.

mattigruener commented:
Pushed a commit that should add some preprocessor directives that fix aforementioned API issues.

andrewkaufman commented:
So I guess the only thing we're waiting for is selection support, right?

andrewkaufman commented:
Just to confirm, what is the maya matrix for which this works?

- Maya 2016
  - [ ] compiles
  - [ ] `scons testMaya` passes
  - [ ] VP1 fully functional
  - [ ] VP2 unsupported
Maya 2017
  - [ ] compiles
  - [ ] `scons testMaya` passes
  - [ ] VP1 fully functional
  - [ ] VP2 full functional (except precision selection issue)
Maya 2018
  - [ ] compiles
  - [ ] `scons testMaya` passes
  - [ ] VP1 fully functional
  - [ ] VP2 full functional

mattigruener commented:
Updated.

99ce076 is a little annoying, but we need a way to make sure that we keep the internal data structure in sync without calling `glScene()`. Once we're ready to rip out the old GL stuff, we'll need to update this bit of code, too.
